packages/core/content-manager/admin/src/preview/utils/previewScript.ts

```diff
@@ -199,8 +236,60 @@ const previewScript = (shouldRun = true) => {
   };
 
   const setupEventHandlers = (highlightManager: HighlightManager) => {
-    // TODO: The listeners for postMessage events will go here
-    return highlightManager.eventListeners;
+    const handleMessage = (event: MessageEvent) => {
+      if (event.data?.type === INTERNAL_EVENTS.STRAPI_FIELD_CHANGE) {
+        const { field, value } = event.data.payload;
+        if (field) {
+          const matchingElements = document.querySelectorAll(`[${SOURCE_ATTRIBUTE}="${field}"]`);
+          matchingElements.forEach((element) => {
+            if (element instanceof HTMLElement) {
+              element.textContent = value || '';
+            }
+          });
+        }
+      } else if (event.data?.type === INTERNAL_EVENTS.STRAPI_FIELD_FOCUS) {
+        const { field } = event.data.payload;
+        if (field) {
+          highlightManager.focusedHighlights.forEach((highlight: HTMLElement) => {
+            highlight.style.outlineColor = 'transparent';
+          });
+          highlightManager.focusedHighlights.length = 0;
+
+          highlightManager.setFocusedField(field);
+          const matchingElements = document.querySelectorAll(`[${SOURCE_ATTRIBUTE}="${field}"]`);
+          matchingElements.forEach((element) => {
+            const highlight =
+              highlightManager.highlights[Array.from(highlightManager.elements).indexOf(element)];
+            if (highlight) {
+              highlight.style.outlineColor = HIGHLIGHT_ACTIVE_COLOR;
+              highlight.style.outlineWidth = '3px';
+              highlightManager.focusedHighlights.push(highlight);
+            }
+          });
+        }
+      } else if (event.data?.type === INTERNAL_EVENTS.STRAPI_FIELD_BLUR) {
+        const { field } = event.data.payload;
+        if (field === highlightManager.getFocusedField()) {
+          highlightManager.focusedHighlights.forEach((highlight: HTMLElement) => {
+            highlight.style.outlineColor = 'transparent';
+            highlight.style.outlineWidth = '2px';
+          });
+          highlightManager.focusedHighlights.length = 0;
+          highlightManager.setFocusedField(null);
+        }
+      }
+    };
```

Comment:
Styling stuff. This is sort of a wall of code that is hard to read. Maybe some spaces could help organize things for the reader? also wondering if the if else if chain is necessary couldn't you just have 3 separate conditions? Did you consider returning early instead of creating new if blocks? like `if (!highlight) return` could help flatten the function a bit. 🤷‍♂️